scalafix-core/src/main/scala/scalafix/patch/Patch.scala

```
@@ -0,0 +1,186 @@
+    def matches(symbol: Symbol): Boolean =
+      if (normalize) symbol.normalized == resolvedFrom
+      else symbol == resolvedFrom
+  }
+
+  @DeriveConfDecoder
+  case class Replace(from: Symbol,
+                     to: Term.Ref,
+                     additionalImports: List[Importer] = Nil,
+                     normalized: Boolean = true)
+      extends TreePatch {
+    require(to.isStableId)
+  }
+  case class RemoveGlobalImport(override val importer: Importer)
+      extends ImportPatch(importer)
+  case class AddGlobalImport(override val importer: Importer)
+      extends ImportPatch(importer)
+}
+
+private[scalafix] object TokenPatch {
+  case class Remove(override val tok: Token) extends TokenPatch(tok, "")
+  case class Add(override val tok: Token,
+                 addLeft: String,
+                 addRight: String,
+                 keepTok: Boolean = true)
+      extends TokenPatch(tok,
+                         s"""$addLeft${if (keepTok) tok else ""}$addRight""")
+
+}
+object Patch {
+  def fromSeq(seq: scala.Seq[Patch]): Patch = seq.foldLeft(empty)(_ + _)
+  def empty: Patch = EmptyPatch
+  def merge(a: TokenPatch, b: TokenPatch): TokenPatch = (a, b) match {
+    case (add1: Add, add2: Add) =>
+      Add(add1.tok,
+          add1.addLeft + add2.addLeft,
+          add1.addRight + add2.addRight,
+          add1.keepTok && add2.keepTok)
+    case (_: Remove, add: Add) => add.copy(keepTok = false)
+    case (add: Add, _: Remove) => add.copy(keepTok = false)
+    case (rem: Remove, _: Remove) => rem
+    case _ =>
+      sys.error(s"""Can't merge token patches:
+                   |1. $a
+                   |2. $b""".stripMargin)
+  }
+  def apply(patch: Patch): String = patch match {
+    case InCtx(p, ctx, Some(mirror)) =>
+      semanticApply(underlying(p))(ctx, mirror)
+    case InCtx(p, ctx, None) =>
+      syntaxApply(ctx, underlying(p).collect { case tp: TokenPatch => tp })
```

TODO: report errors on TreePatches instead of silently ignoring them